drivers/counter/counter_gd32_timer.c

```diff
@@ -0,0 +1,538 @@
+};
+
+static uint32_t get_autoreload_value(const struct device *dev)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	return TIMER_CAR(config->reg);
+}
+
+static void set_autoreload_value(const struct device *dev, uint32_t value)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	TIMER_CAR(config->reg) = value;
+}
+
+static uint32_t get_counter(const struct device *dev)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	return TIMER_CNT(config->reg);
+}
+
+static void set_counter(const struct device *dev, uint32_t value)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	TIMER_CNT(config->reg) = value;
+}
+
+static void set_software_event_gen(const struct device *dev, uint8_t evt)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	TIMER_SWEVG(config->reg) |= evt;
+}
+
+static void set_prescaler(const struct device *dev, uint16_t prescaler)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	TIMER_PSC(config->reg) = prescaler;
+}
+
+static void set_compare_value(const struct device *dev, uint16_t chan,
+			      uint32_t now)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	switch (chan) {
+	case 0:
+		TIMER_CH0CV(config->reg) = now;
+		break;
+	case 1:
+		TIMER_CH1CV(config->reg) = now;
+		break;
+	case 2:
+		TIMER_CH2CV(config->reg) = now;
+		break;
+	case 3:
+		TIMER_CH3CV(config->reg) = now;
+		break;
+	}
+}
+
+static void interrupt_enable(const struct device *dev, uint32_t interrupt)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	TIMER_DMAINTEN(config->reg) |= interrupt;
+}
+
+static void interrupt_disable(const struct device *dev, uint32_t interrupt)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	TIMER_DMAINTEN(config->reg) &= ~interrupt;
+}
+
+static uint32_t interrupt_flag_get(const struct device *dev, uint32_t interrupt)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	return (TIMER_DMAINTEN(config->reg) & TIMER_INTF(config->reg) &
+		interrupt);
+}
+
+static void interrupt_flag_clear(const struct device *dev, uint32_t interrupt)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	TIMER_INTF(config->reg) &= ~interrupt;
+}
+
+static int counter_gd32_timer_start(const struct device *dev)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	TIMER_CTL0(config->reg) |= (uint32_t)TIMER_CTL0_CEN;
+
+	return 0;
+}
+
+static int counter_gd32_timer_stop(const struct device *dev)
+{
+	const struct counter_gd32_config *config = dev->config;
+
+	TIMER_CTL0(config->reg) &= ~(uint32_t)TIMER_CTL0_CEN;
+
+	return 0;
+}
+
+static int counter_gd32_timer_get_value(const struct device *dev,
+					uint32_t *ticks)
+{
+	*ticks = get_counter(dev);
+
+	return 0;
+}
+
+static uint32_t counter_gd32_timer_get_top_value(const struct device *dev)
+{
+	return get_autoreload_value(dev);
+}
+
+/* Return true if value equals 2^n - 1 */
+static inline bool is_bit_mask(uint32_t val)
```

Comment:
there is already `is_power_of_two` in `sys/util.h` and i couldn't find function like this one there. It would be worth to move it there next to `is_power_of_two`.

Reply:
I will work on [XXX ](https://github.com/zephyrproject-rtos/zephyr/pull/52956) after this PR merged.